PlaceholderDifferenceEvaluator throws ClassCastException when using XMLSchema-instance (xsi)

When using polymorphism in XML using XMLSchema-instance the `PlaceholderDifferenceEvaluator` runs into an exception, when the `xsi:type` attribute has a different value:

**Exception:**
`java.lang.ClassCastException: class javax.xml.namespace.QName cannot be cast to class java.lang.String`

**JUnit test case that verifies the problem:**
```
    @Test
    void placeholderDifferenceEvaluatorWithXmlInstance() {
        var expectation = """
                <?xml version="1.0" encoding="UTF-8" standalone="yes"?>
                <myns:root xmlns:myns="http://example.com/myns" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
                  <myns:property xsi:type="myns:firstType">
                    value
                  </myns:property>
                </myns:root>
                """;

        var xml = """
                <?xml version="1.0" encoding="UTF-8" standalone="yes"?>
                <myns:root xmlns:myns="http://example.com/myns" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
                  <myns:property xsi:type="myns:secondType">
                    value
                  </myns:property>
                </myns:root>
                """;

        assertDoesNotThrow(() -> {
            DiffBuilder
                    .compare(Input.fromString(expectation))
                    .withTest(Input.fromString(xml))
                    .ignoreWhitespace()
                    .withDifferenceEvaluator(new PlaceholderDifferenceEvaluator())
                    .build();
        });
    }
```

**Stacktrace:**
```
Caused by: java.lang.ClassCastException: class javax.xml.namespace.QName cannot be cast to class java.lang.String (javax.xml.namespace.QName is in module java.xml of loader 'bootstrap'; java.lang.String is in module java.base of loader 'bootstrap')
	at org.xmlunit.placeholder.PlaceholderDifferenceEvaluator.evaluate(PlaceholderDifferenceEvaluator.java:210)
	at org.xmlunit.diff.AbstractDifferenceEngine.compare(AbstractDifferenceEngine.java:187)
	at org.xmlunit.diff.AbstractDifferenceEngine$ComparisonState$2.apply(AbstractDifferenceEngine.java:286)
	at org.xmlunit.diff.AbstractDifferenceEngine$ComparisonState.andThen(AbstractDifferenceEngine.java:248)
	at org.xmlunit.diff.AbstractDifferenceEngine$ComparisonState.andIfTrueThen(AbstractDifferenceEngine.java:259)
	at org.xmlunit.diff.AbstractDifferenceEngine$ComparisonState.andIfTrueThen(AbstractDifferenceEngine.java:283)
	at org.xmlunit.diff.DOMDifferenceEngine.compareXsiType(DOMDifferenceEngine.java:753)
	at org.xmlunit.diff.DOMDifferenceEngine.access$800(DOMDifferenceEngine.java:47)
	at org.xmlunit.diff.DOMDifferenceEngine$7.apply(DOMDifferenceEngine.java:411)
	at org.xmlunit.diff.AbstractDifferenceEngine$ComparisonState.andThen(AbstractDifferenceEngine.java:248)
	at org.xmlunit.diff.DOMDifferenceEngine.compareElementAttributes(DOMDifferenceEngine.java:408)
	at org.xmlunit.diff.DOMDifferenceEngine.access$200(DOMDifferenceEngine.java:47)
	at org.xmlunit.diff.DOMDifferenceEngine$6.apply(DOMDifferenceEngine.java:381)
	at org.xmlunit.diff.AbstractDifferenceEngine$ComparisonState.andThen(AbstractDifferenceEngine.java:248)
	at org.xmlunit.diff.DOMDifferenceEngine.compareElements(DOMDifferenceEngine.java:378)
	at org.xmlunit.diff.DOMDifferenceEngine.nodeTypeSpecificComparison(DOMDifferenceEngine.java:216)
	at org.xmlunit.diff.DOMDifferenceEngine.access$000(DOMDifferenceEngine.java:47)
	at org.xmlunit.diff.DOMDifferenceEngine$2.apply(DOMDifferenceEngine.java:176)
	at org.xmlunit.diff.AbstractDifferenceEngine$ComparisonState.andThen(AbstractDifferenceEngine.java:248)
	at org.xmlunit.diff.DOMDifferenceEngine.compareNodes(DOMDifferenceEngine.java:173)
	at org.xmlunit.diff.DOMDifferenceEngine$8.apply(DOMDifferenceEngine.java:619)
	at org.xmlunit.diff.AbstractDifferenceEngine$ComparisonState.andThen(AbstractDifferenceEngine.java:248)
	at org.xmlunit.diff.DOMDifferenceEngine.compareNodeLists(DOMDifferenceEngine.java:616)
	at org.xmlunit.diff.DOMDifferenceEngine.access$100(DOMDifferenceEngine.java:47)
	at org.xmlunit.diff.DOMDifferenceEngine$3.apply(DOMDifferenceEngine.java:260)
	at org.xmlunit.diff.AbstractDifferenceEngine$ComparisonState.andThen(AbstractDifferenceEngine.java:248)
	at org.xmlunit.diff.AbstractDifferenceEngine$ComparisonState.andIfTrueThen(AbstractDifferenceEngine.java:259)
	at org.xmlunit.diff.DOMDifferenceEngine.compareNodes(DOMDifferenceEngine.java:181)
	at org.xmlunit.diff.DOMDifferenceEngine$8.apply(DOMDifferenceEngine.java:619)
	at org.xmlunit.diff.AbstractDifferenceEngine$ComparisonState.andThen(AbstractDifferenceEngine.java:248)
	at org.xmlunit.diff.DOMDifferenceEngine.compareNodeLists(DOMDifferenceEngine.java:616)
	at org.xmlunit.diff.DOMDifferenceEngine.access$100(DOMDifferenceEngine.java:47)
	at org.xmlunit.diff.DOMDifferenceEngine$3.apply(DOMDifferenceEngine.java:260)
	at org.xmlunit.diff.AbstractDifferenceEngine$ComparisonState.andThen(AbstractDifferenceEngine.java:248)
	at org.xmlunit.diff.AbstractDifferenceEngine$ComparisonState.andIfTrueThen(AbstractDifferenceEngine.java:259)
	at org.xmlunit.diff.DOMDifferenceEngine.compareNodes(DOMDifferenceEngine.java:181)
	at org.xmlunit.diff.DOMDifferenceEngine.compare(DOMDifferenceEngine.java:117)
	... 5 more

```

--- a/RELEASE_NOTES.md
+++ b/RELEASE_NOTES.md
@@ -2,6 +2,13 @@
 
 ## XMLUnit for Java 2.10.1 - /not released, yet/
 
+* placeholders can now also be used inside of the local part of `xsi:type` attributes.
+  PR [#293](https://github.com/xmlunit/xmlunit/pull/293)
+
+* PlaceholderDifferenceEvaluator would cause ClassCastException for documents with
+  differences in `xsi:type` attributes.
+  Issue [#276](https://github.com/xmlunit/xmlunit/issues/276)
+
 * updated a bunch of Maven plugins, in particular we now create CylcloneDX files using
   version 1.6 of the schema.
   PR [#292](https://github.com/xmlunit/xmlunit/pull/292)

--- a/xmlunit-placeholders/src/main/java/org/xmlunit/placeholder/PlaceholderDifferenceEvaluator.java
+++ b/xmlunit-placeholders/src/main/java/org/xmlunit/placeholder/PlaceholderDifferenceEvaluator.java
@@ -19,6 +19,7 @@ import java.util.Map;
 import java.util.ServiceLoader;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
+import javax.xml.XMLConstants;
 import javax.xml.namespace.QName;
 import org.w3c.dom.Node;
 import org.xmlunit.diff.Comparison;
@@ -212,6 +213,13 @@ public class PlaceholderDifferenceEvaluator implements DifferenceEvaluator {
             return evaluateConsideringPlaceholders((String) controlDetails.getValue(),
                 (String) testDetails.getValue(), outcome);
 
+        // comparing QName content of xsi:type attributes
+        } else if (comparison.getType() == ComparisonType.ATTR_VALUE
+                   && controlDetails.getValue() instanceof QName
+                   && testDetails.getValue() instanceof QName) {
+            return evaluateConsideringPlaceholders((QName) controlDetails.getValue(),
+                (QName) testDetails.getValue(), outcome);
+
         // "test document has no attribute but control document has"
         } else if (isMissingAttributeDifference(comparison)) {
             return evaluateMissingAttributeConsideringPlaceholders(comparison, outcome);
@@ -303,6 +311,16 @@ public class PlaceholderDifferenceEvaluator implements DifferenceEvaluator {
             return outcome;
         }
         return ComparisonResult.EQUAL;
+    }
+
+    private ComparisonResult evaluateConsideringPlaceholders(QName controlQName, QName testQName,
+        ComparisonResult outcome) {
+        if (controlQName.getNamespaceURI().equals(testQName.getNamespaceURI())) {
+            return evaluateConsideringPlaceholders(controlQName.getLocalPart(), testQName.getLocalPart(),
+                                                   outcome);
+        }
+
+        return outcome;
     }
 
     private ComparisonResult evaluateConsideringPlaceholders(String controlText, String testText,

--- a/xmlunit-placeholders/src/main/java/org/xmlunit/placeholder/PlaceholderDifferenceEvaluator.java
+++ b/xmlunit-placeholders/src/main/java/org/xmlunit/placeholder/PlaceholderDifferenceEvaluator.java
@@ -206,7 +206,9 @@ public class PlaceholderDifferenceEvaluator implements DifferenceEvaluator {
             return evaluateConsideringPlaceholders(controlTarget.getNodeValue(), testTarget.getNodeValue(), outcome);
 
         // comparing textual content of attributes
-        } else if (comparison.getType() == ComparisonType.ATTR_VALUE) {
+        } else if (comparison.getType() == ComparisonType.ATTR_VALUE
+                   && controlDetails.getValue() instanceof String
+                   && testDetails.getValue() instanceof String) {
             return evaluateConsideringPlaceholders((String) controlDetails.getValue(),
                 (String) testDetails.getValue(), outcome);
 

--- a/xmlunit-placeholders/src/main/java/org/xmlunit/placeholder/package-info.java
+++ b/xmlunit-placeholders/src/main/java/org/xmlunit/placeholder/package-info.java
@@ -78,6 +78,11 @@
  * org.xmlunit.diff.DifferenceEvaluator} and add it to the builder
  * yourself.</p>
  *
+ * <p>Placeholder sequences must appear as values inside of attribute values or nested textual content of elements -
+ * inside the control document. As a special case they may also appear as local part of a tye name of xsi:type
+ * attributes of control documents - i.e. <code>xsi:type="prefix:${xmlunit:ignore}"</code>. The namepace URIs of the
+ * compared types must match with the test document. Support for xsi:types has been added with XMLUnit 2.10.1.</p>
+ *
  * @since 2.6.0
  */
 package org.xmlunit.placeholder;

--- a/xmlunit-placeholders/src/test/java/org/xmlunit/placeholder/PlaceholderDifferenceEvaluatorTest.java
+++ b/xmlunit-placeholders/src/test/java/org/xmlunit/placeholder/PlaceholderDifferenceEvaluatorTest.java
@@ -517,6 +517,23 @@ public class PlaceholderDifferenceEvaluatorTest {
     }
 
     @Test
+    public void cantIgnoreXsiNamespaceDifference() {
+        String control = "<element"
+            + " xmlns:myns=\"https://example.org/some-other-ns\""
+            + " xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\""
+            + " xsi:type=\"myns:${xmlunit.ignore}\" />";
+        String test = "<element"
+            + " xmlns:myns=\"https://example.org/some-ns\""
+            + " xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\""
+            + " xsi:type=\"myns:some-other-type\" />";
+
+        Diff diff = DiffBuilder.compare(control).withTest(test)
+            .withDifferenceEvaluator(new PlaceholderDifferenceEvaluator()).build();
+
+        assertTrue(diff.hasDifferences());
+    }
+
+    @Test
     public void canCompareXsiTypeWithhRegex() {
         String control = "<element"
             + " xmlns:myns=\"https://example.org/some-ns\""

--- a/xmlunit-placeholders/src/test/java/org/xmlunit/placeholder/PlaceholderDifferenceEvaluatorTest.java
+++ b/xmlunit-placeholders/src/test/java/org/xmlunit/placeholder/PlaceholderDifferenceEvaluatorTest.java
@@ -498,4 +498,38 @@ public class PlaceholderDifferenceEvaluatorTest {
         }
         assertEquals(1, count);
     }
+
+    @Test
+    public void canIgnoreXsiTypeDifference() {
+        String control = "<element"
+            + " xmlns:myns=\"https://example.org/some-ns\""
+            + " xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\""
+            + " xsi:type=\"myns:${xmlunit.ignore}\" />";
+        String test = "<element"
+            + " xmlns:myns=\"https://example.org/some-ns\""
+            + " xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\""
+            + " xsi:type=\"myns:some-other-type\" />";
+
+        Diff diff = DiffBuilder.compare(control).withTest(test)
+            .withDifferenceEvaluator(new PlaceholderDifferenceEvaluator()).build();
+
+        assertFalse(diff.hasDifferences());
+    }
+
+    @Test
+    public void canCompareXsiTypeWithhRegex() {
+        String control = "<element"
+            + " xmlns:myns=\"https://example.org/some-ns\""
+            + " xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\""
+            + " xsi:type=\"myns:${xmlunit.matchesRegex(.*-type)}\" />";
+        String test = "<element"
+            + " xmlns:myns=\"https://example.org/some-ns\""
+            + " xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\""
+            + " xsi:type=\"myns:some-other-type\" />";
+
+        Diff diff = DiffBuilder.compare(control).withTest(test)
+            .withDifferenceEvaluator(new PlaceholderDifferenceEvaluator()).build();
+
+        assertFalse(diff.hasDifferences());
+    }
 }

--- a/xmlunit-placeholders/src/test/java/org/xmlunit/placeholder/PlaceholderDifferenceEvaluatorTest.java
+++ b/xmlunit-placeholders/src/test/java/org/xmlunit/placeholder/PlaceholderDifferenceEvaluatorTest.java
@@ -472,4 +472,30 @@ public class PlaceholderDifferenceEvaluatorTest {
         assertFalse(diff.hasDifferences());
     }
 
+    @Test
+    public void canCompareDocmentsWithXsiTypes() {
+        String control = "<element"
+            + " xmlns:myns=\"https://example.org/some-ns\""
+            + " xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\""
+            + " xsi:type=\"myns:some-type\" />";
+        String test = "<element"
+            + " xmlns:myns=\"https://example.org/some-ns\""
+            + " xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\""
+            + " xsi:type=\"myns:some-other-type\" />";
+
+        Diff diff = DiffBuilder.compare(control).withTest(test)
+            .withDifferenceEvaluator(new PlaceholderDifferenceEvaluator()).build();
+
+        assertTrue(diff.hasDifferences());
+        int count = 0;
+        Iterator it = diff.getDifferences().iterator();
+        while (it.hasNext()) {
+            count++;
+            Difference difference = (Difference) it.next();
+            assertEquals(ComparisonResult.DIFFERENT, difference.getResult());
+            assertEquals(ComparisonType.ATTR_VALUE,
+                         difference.getComparison().getType());
+        }
+        assertEquals(1, count);
+    }
 }

